C# V7: Generalized async return types (#556)
dotnet/csharpstandard: standard/classes.md

It is a compile-time error for the formal parameter list of an async function to specify any `ref` or `out` parameters.

The *return_type* of an async method shall be either `void` or a ***task type***. For an async method that returns a value, a task type shall be generic. For an async method that does not return a value, a task type shall not be generic. Such types are referred to in this specification as `«TaskType»<T>` and `«TaskType»`, respectively. (The Standard library types `System.Threading.Tasks.Task<T>` and types constructed from `System.Threading.Tasks.Task` are task types.)

> A task type shall be a class or struct type that is associated with a ***builder type*** via the attribute `System.Runtime.CompilerServices.AsyncMethodBuilder`. Such types are referred to in this specification as `«TaskBuilderType»<T>` and `«TaskBuilderType»`.

Reviewer: Should we call this task builder type instead of builder type? It feels too general at the moment.

An async method returning a task type is said to be ***task-returning***.

The exact definition of the task types is implementation-defined, but from the language’s point of view, a task type is in one of the states *incomplete*, *succeeded* or *faulted*. A *faulted* task records a pertinent exception. A *succeeded* `«TaskType»<T>` records a result of type `T`. Task types are awaitable, and tasks can therefore be the operands of await expressions ([§11.8.8](expressions.md#1188-await-expressions)).

> > *Example*: The task type `MyTask<T>` is associated to the builder type `MyTaskMethodBuilder<T>` and the awaiter type `Awaiter<T>`:

Reviewer: ```suggestion
> *Example*: The task type `MyTask<T>` is associated with the builder type `MyTaskMethodBuilder<T>` and the awaiter type `Awaiter<T>`:
```

>
> <!-- Example: {template:"standalone-lib-without-using", name:"AsyncFunctions1", replaceEllipsis:true, customEllipsisReplacements: ["return new Awaiter<T>();", "", "return default(T);"], additionalFiles:["MyTaskMethodBuilderT.cs"]} -->
> ```csharp
> using System.Runtime.CompilerServices; 
> [AsyncMethodBuilder(typeof(MyTaskMethodBuilder<>))]
> class MyTask<T>
> {
>     public Awaiter<T> GetAwaiter() { ... }
> }
>
> class Awaiter<T> : INotifyCompletion
> {
>     public void OnCompleted(Action completion) { ... }
>     public bool IsCompleted { get; }
>     public T GetResult() { ... }
> }
> ```
>
> *end example*

> A builder type is a class or struct type that corresponds to a specific task type. A builder type can have at most one type parameter and cannot be nested in a generic type. A builder type shall have the following public members (for non-generic builder types, `SetResult` has no parameters):

Reviewer 2: Per previous discussion:

```suggestion
A builder type is a class or struct type that corresponds to a specific task type. A builder type can have at most one type parameter and cannot be nested in a generic type. A builder type shall have the following accessible members (for non-generic builder types, `SetResult` has no parameters):
```

> A builder type is a class or struct type that corresponds to a specific task type. A builder type can have at most one type parameter and cannot not be nested in a generic type. A builder type shall have the following public methods (for non-generic builder types, `SetResult` has no parameters):

Reviewer: ```suggestion
A builder type is a class or struct type that corresponds to a specific task type. A builder type can have at most one type parameter and cannot be nested in a generic type. A builder type shall have the following public members (for non-generic builder types, `SetResult` has no parameters):
```

(Changes methods to members due to the `Task` property.)

Reviewer: Do the members actually need to be public, if the task builder type is only being used in the same assembly as its declaration? (I'd expect them to only have to be accessible at the point of use.)

Reviewer 2: Correct. They must be accessible.

> A builder type is a class or struct type that corresponds to a specific task type. A builder type can have at most one type parameter and cannot not be nested in a generic type. A builder type shall have the following public methods (for non-generic builder types, `SetResult` has no parameters):

Reviewer: It feels like this should be in "Task-type builder pattern" section (possibly with a link from here, although after moving there isn't much between this location and where it will be anyway).

Reviewer 2: I like it. Made that change.

    public void AwaitUnsafeOnCompleted<TAwaiter, TStateMachine>(
        ref TAwaiter awaiter, ref TStateMachine stateMachine)
        where TAwaiter : ICriticalNotifyCompletion
        where TStateMachine : IAsyncStateMachine;
    public «TaskType»<T> Task { get; }
}
```
  
An async function has the ability to suspend evaluation by means of await expressions ([§11.8.8](expressions.md#1188-await-expressions)) in its body. Evaluation may later be resumed at the point of the suspending await expression by means of a ***resumption delegate***. The resumption delegate is of type `System.Action`, and when it is invoked, evaluation of the async function invocation will resume from the await expression where it left off. The ***current caller*** of an async function invocation is the original caller if the function invocation has never been suspended or the most recent caller of the resumption delegate otherwise.

### §task-builder-pattern Task-type builder pattern

The compiler generates code that uses the «TaskBuilderType» to implement the semantics of suspending and resuming the evaluation of the async function. The uses the «TaskBuilderType» as follows:

> - `Builder.Create()` is invoked to create an instance of the «TaskBuilderType», named `builder` in this list.

Reviewer: ```suggestion
- `«TaskBuilderType».Create()` is invoked to create an instance of the «TaskBuilderType», named `builder` in this list.
```

  - The builder must call `stateMachine.MoveNext()` either in `Start()` or after `Start()` has returned to advance the state machine.
- After `Start()` returns, the `async` method invokes `builder.Task` for the task to return from the async method.
- Each call to `stateMachine.MoveNext()` will advance the state machine.
- If the state machine completes successfully, `builder.SetResult()` is called, with the method return value, if any.
- Otherwise, if an exception, `e` is thrown in the state machine, `builder.SetException(e)` is called.
- If the state machine reaches an `await expr` expression, `expr.GetAwaiter()` is invoked.
- If the awaiter implements `ICriticalNotifyCompletion` and `IsCompleted` is false, the state machine invokes `builder.AwaitUnsafeOnCompleted(ref awaiter, ref stateMachine)`.
  - `AwaitUnsafeOnCompleted()` should call `awaiter.UnsafeOnCompleted(action)` with an `Action` that calls `stateMachine.MoveNext()` when the awaiter completes.
- Otherwise, the state machine invokes `builder.AwaitOnCompleted(ref awaiter, ref stateMachine)`.

>   - `AwaitOnCompleted()` should call `awaiter.OnCompleted(action)`.

Reviewer: ```suggestion
  - `AwaitOnCompleted()` should call `awaiter.OnCompleted(action)` with an `Action` that calls `stateMachine.MoveNext()` when the awaiter completes.
```

Reviewer: (For symmetry with the unsafe version.)

- `SetStateMachine(IAsyncStateMachine)` may be called by the compiler-generated `IAsyncStateMachine` implementation to identify the instance of the builder associated with a state machine instance, particularly for cases where the state machine is implemented as a value type.
  - If the builder calls `stateMachine.SetStateMachine(stateMachine)`, the `stateMachine` will call `builder.SetStateMachine(stateMachine)` on the *builder instance associated with* `stateMachine`.

> > *Note:* The generated code is equivalent to the code generated for async methods that return `Task`, `Task<T>`, or `void`. The difference is, for those well known types, the *builder types* are also known to the compiler. In other words, the semantics for a *builder-type* should match the semantics of the known builder types for `Task`, `Task<T>`, or `void` *end note*

Reviewer: Is this actually an implementation detail? (It would be reasonable to have an entirely different pattern for the implementation-specific known builder types.) I guess "the semantics" should be the same in the end, but there could be quite different members available.

Reviewer 2: I tend to agree. As a note, it's not normative. Do you think we should remove it?

Reviewer: I think that's best at least for now, yes.

> > *Note:* The generated code is equivalent to the code generated for async methods that return `Task`, `Task<T>`, or `void`. The difference is, for those well known types, the *task builder types* are also known to the compiler. In other words, the semantics for a *task-builder-type* should match the semantics of the known task builder types for `Task`, `Task<T>`, or `void` *end note*

Reviewer 2: Remove note per feedback:

```suggestion
```
